### ISSUE TITLE
question: debugging zone.js performance issues 🐌  🌩️

### ISSUE THREAD

**Issue by samccone** [angular/zone.js#724]
I have recently been debugging angular 2 apps who have been running into performance problems rooted in what seems to be zone.ts (I would share screenshots but they are currently not public things, and yes I acknowledge this makes my question harder to answer directly).

my observations have been that I have a lot of call stacks that terminating or originate in the following root functions:

scheduleMicroTask
scheduleMacroTask
scheduleQueueDrain

or in some other cases timer

--

My normal approach for debugging perf issues follows the ideas of M.P.D.I.A
measure, profile, diagnose, identify, attempt 

as @TheLarkInn so beautifully captured 
![image](https://cloud.githubusercontent.com/assets/883126/24736788/155f9900-1a3f-11e7-955b-4f695df75693.png)

In the real world this looks like:

1. record a timeline
2. look at a segment of the timeline using the bottom up profiler
3. identify the heaviest stacks
4. identify what is starting the stacks to happen (a callback? a user action? a loop?)
5. change what is causing the work and see if there is an impact

But with zone.ts code, i see very very long stacks 99% of which is inside of zone code (counting self time) and very little of the self time in userland code, making my process hard to follow.

So here is my question, how does one go about profiling code that is using zone? Is there a way to disable zone to see if it is the root issue? Is there a special perf technique for debugging zone code that requires knowledge of how zone is implemented?

Thanks in advance!

**Comment by Nilos** [comment]
I had to find out recently that zone.js groups all Promise.then callbacks together and executes them in series, which might be one of the reasons performance problems occur (or are hard to debug as they are now all hidden behind one big micro task). See #743

**Comment by SamVerschueren** [comment]
I'm experiencing the same struggles. Zone.js is firing every 50ms and seems to be stuck in a loop. I can't detect the root cause because of this.

@samccone Did you find any solution to this in the mean time?

**Comment by instantaphex** [comment]
@samccone @SamVerschueren Any luck?  I'm dealing with the same issue here.

**Comment by mleibman** [comment]
What is your definition of "userland code"? Zone.js wraps the scheduling and execution of tasks, so it makes sense that most calls would originate in Zone.js. If you're looking at a stack trace with async gaps in it, you will also see each stacktrace ending in the async gap terminate in Zone.js as well, since that's where the async task gets scheduled.

@SamVerschueren Debugging this should be pretty easy. Set a breakpoint on whatever is executing every 50ms and look at Zone.currentTask.source.
